# DecodeLabs Internship — Project 1
## Data Cleaning and Preparation

### Objective

The objective of this project is to clean and prepare an e-commerce order dataset for further analysis by identifying and handling missing values, duplicate records, data-format issues, and other data-quality problems.

### Dataset Overview

- **Records:** 1,200
- **Columns:** 14
- **File type:** Excel (.xlsx)

### Tools Used

- Python
- Pandas
- Google Colab

### Data Cleaning Tasks

The dataset was examined for:

1. Missing values
2. Duplicate records
3. Duplicate Order IDs
4. Date formatting
5. Numerical data consistency
6. Text/categorical consistency
7. Identifier validity

### Final Validation

The cleaned dataset was validated to ensure:

- Zero missing values
- Zero duplicate rows
- Zero duplicate Order IDs
- Zero duplicate Tracking Numbers
- Correct date data type
- Correct TotalPrice calculations

### Output

The final cleaned dataset was exported as:

`Cleaned_Dataset.xlsx`

In [4]:
from google.colab import files
uploaded = files.upload()

Saving Dataset for Data Analytics.xlsx to Dataset for Data Analytics (1).xlsx


Step 1- Data Loading and Initial Inspection

1.1 - Import required libraries

In [5]:
import pandas as pd

1.2 -  Load dataset

In [6]:
df = pd.read_excel("Dataset for Data Analytics.xlsx")

1.3 - View dataset

In [7]:
df.head(15)

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04
5,ORD200005,2023-10-23,C37249,Phone,2,245.86,934 Main St,Credit Card,Shipped,TRK72976927,4,SAVE10,Instagram,491.72
6,ORD200006,2025-06-17,C83492,Laptop,1,664.42,986 Main St,Gift Card,Returned,TRK96417362,6,SAVE10,Facebook,664.42
7,ORD200007,2023-05-12,C41460,Monitor,5,149.55,706 Main St,Cash,Shipped,TRK78809193,9,FREESHIP,Facebook,747.75
8,ORD200008,2025-04-02,C26817,Phone,2,134.28,904 Main St,Gift Card,Cancelled,TRK61042692,2,NaN,Email,268.56
9,ORD200009,2023-11-21,C31946,Desk,4,509.38,102 Main St,Credit Card,Shipped,TRK33478363,6,SAVE10,Google,2037.52


1.4 - Understand the dataset

In [8]:
df.shape #Check number of rows and columns

(1200, 14)

In [9]:
df.columns #See list of all column names

Index(['OrderID', 'Date', 'CustomerID', 'Product', 'Quantity', 'UnitPrice',
       'ShippingAddress', 'PaymentMethod', 'OrderStatus', 'TrackingNumber',
       'ItemsInCart', 'CouponCode', 'ReferralSource', 'TotalPrice'],
      dtype='object')

In [10]:
df.info() #Check datatype and basic info. of all columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   OrderID          1200 non-null   object        
 1   Date             1200 non-null   datetime64[ns]
 2   CustomerID       1200 non-null   object        
 3   Product          1200 non-null   object        
 4   Quantity         1200 non-null   int64         
 5   UnitPrice        1200 non-null   float64       
 6   ShippingAddress  1200 non-null   object        
 7   PaymentMethod    1200 non-null   object        
 8   OrderStatus      1200 non-null   object        
 9   TrackingNumber   1200 non-null   object        
 10  ItemsInCart      1200 non-null   int64         
 11  CouponCode       891 non-null    object        
 12  ReferralSource   1200 non-null   object        
 13  TotalPrice       1200 non-null   float64       
dtypes: datetime64[ns](1), float64(2), int64(

2. Missing Value Analysis

In [11]:
df.isnull().sum() #Check for missing values

,0
OrderID,0
Date,0
CustomerID,0
Product,0
Quantity,0
UnitPrice,0
ShippingAddress,0
PaymentMethod,0
OrderStatus,0
TrackingNumber,0


In [12]:
(df.isnull().sum() / len(df) * 100).round(2) #percentage of missing values

,0
OrderID,0.00
Date,0.00
CustomerID,0.00
Product,0.00
Quantity,0.00
UnitPrice,0.00
ShippingAddress,0.00
PaymentMethod,0.00
OrderStatus,0.00
TrackingNumber,0.00


2.1 - Inspect the CouponCode column

In [13]:
df['CouponCode'].value_counts(dropna=False)

,count
CouponCode,
FREESHIP,313
NaN,309
WINTER15,292
SAVE10,286


In [14]:
df['CouponCode'].unique()

array(['SAVE10', 'FREESHIP', nan, 'WINTER15'], dtype=object)

In [15]:
df[df['CouponCode'].isna()].head(12) #see some records with missing coupon

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
8,ORD200008,2025-04-02,C26817,Phone,2,134.28,904 Main St,Gift Card,Cancelled,TRK61042692,2,NaN,Email,268.56
15,ORD200015,2023-07-17,C39416,Printer,1,473.96,942 Main St,Cash,Delivered,TRK54930938,3,NaN,Google,473.96
17,ORD200017,2024-03-02,C60672,Tablet,1,423.40,382 Main St,Debit Card,Shipped,TRK11972116,3,NaN,Referral,423.40
21,ORD200021,2024-11-17,C99023,Monitor,4,342.95,569 Main St,Cash,Delivered,TRK52646983,6,NaN,Google,1371.80
26,ORD200026,2024-03-06,C21944,Laptop,5,59.55,296 Main St,Online,Delivered,TRK80276812,6,NaN,Referral,297.75
28,ORD200028,2023-11-21,C59823,Phone,4,160.07,568 Main St,Online,Pending,TRK65895539,5,NaN,Email,640.28
31,ORD200031,2023-06-30,C26445,Chair,3,500.08,996 Main St,Debit Card,Returned,TRK31229981,6,NaN,Facebook,1500.24
32,ORD200032,2023-05-01,C12388,Tablet,5,536.72,830 Main St,Cash,Delivered,TRK99261395,10,NaN,Email,2683.60
36,ORD200036,2025-05-08,C59337,Laptop,2,534.43,465 Main St,Debit Card,Pending,TRK50961250,5,NaN,Facebook,1068.86
42,ORD200042,2024-05-06,C28712,Printer,2,151.66,980 Main St,Cash,Cancelled,TRK40772558,6,NaN,Instagram,303.32


2.2 - Handling the missing values in the CouponCode column

In [18]:
df['CouponCode'] = df['CouponCode'].fillna('Unknown')  #fill missing coupon values with 'unknown' because the dataset itself doesn't explicitly tell us that NaN means "the customer did not use a coupon."

In [72]:
print(df['CouponCode'].isnull().sum()) #Verify that there are no missing values left

0


In [24]:
df['CouponCode'].value_counts()

,count
CouponCode,
FREESHIP,313
Unknown,309
WINTER15,292
SAVE10,286


The dataset was checked for missing values using Pandas. The analysis identified 309 missing values in the `CouponCode` column. All other columns contained no missing values.

The missing `CouponCode` values were investigated by examining the affected records and the unique values in the column. Since the dataset did not provide enough information to determine whether a missing coupon represented "no coupon" or unavailable information, the missing values were replaced with `Unknown`.

After the replacement, the `CouponCode` column was verified to contain zero missing values.

3. Duplicate Analysis

3.1 - Check for duplicate rows

In [26]:
print(df.duplicated().sum())  #check whether the dataset contains completely duplicated records.

0


3.2 - Check duplicate Order IDs

In [27]:
print(df['OrderID'].duplicated().sum())

0


In [28]:
df[df['OrderID'].duplicated(keep=False)]

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice


The dataset was examined for completely duplicated rows and duplicate `OrderID` values.

The analysis found:

- Duplicate rows: 0
- Duplicate `OrderID` values: 0

Therefore, no records were removed due to duplication.

4. Check the data types

In [29]:
df.dtypes

,0
OrderID,object
Date,datetime64[ns]
CustomerID,object
Product,object
Quantity,int64
UnitPrice,float64
ShippingAddress,object
PaymentMethod,object
OrderStatus,object
TrackingNumber,object


4.1 - Verify the Date values

In [31]:
df['Date'].head(20) #How the dates are represented.

,Date
0,2023-01-04
1,2024-08-23
2,2024-02-27
3,2023-10-15
4,2025-05-08
5,2023-10-23
6,2025-06-17
7,2023-05-12
8,2025-04-02
9,2023-11-21


In [35]:
print("The order dates lie between", df['Date'].min(),"and", df['Date'].max()) #The date range in the dataset.

The order dates lie between 2023-01-01 00:00:00 and 2025-06-30 00:00:00


In [75]:
print("Number of missing date values:" ,(df['Date'].isna().sum()))

Number of missing date values: 0


The `Date` column was examined to verify that the order dates were stored in a consistent and appropriate date format.

The column was found to have the `datetime64[ns]` data type. The dates were consistently represented in the `YYYY-MM-DD` format.

The order dates ranged from `2023-01-01` to `2025-06-30`, and no missing dates were identified.

Since the date format and data type were already correct, no changes were required.

5 - Check the numerical columns

In [36]:
df[['Quantity', 'UnitPrice', 'ItemsInCart', 'TotalPrice']].describe()

,Quantity,UnitPrice,ItemsInCart,TotalPrice
count,1200.000000,1200.000000,1200.000000,1200.000000
mean,2.945833,356.412750,5.485000,1053.968300
std,1.407557,197.177146,2.281983,819.856558
min,1.000000,11.390000,1.000000,11.390000
25%,2.000000,186.062500,4.000000,410.520000
50%,3.000000,364.210000,5.000000,823.615000
75%,4.000000,521.570000,7.000000,1578.475000
max,5.000000,699.930000,10.000000,3456.400000


The numerical columns `Quantity`, `UnitPrice`, `ItemsInCart`, and `TotalPrice` were analyzed using descriptive statistics.

The values were checked for missing, zero, and negative values.

The following observations were made:

- `Quantity` ranged from 1 to 5.
- `UnitPrice` ranged from 11.39 to 699.93.
- `ItemsInCart` ranged from 1 to 10.
- `TotalPrice` ranged from 11.39 to 3456.40.

No invalid negative or zero values were identified in these columns.

5.1 - Check whether TotalPrice is calculated correctly

In [39]:
print("Number of TotalPrice that are not calculated correctlly: ",(df['Quantity'] * df['UnitPrice'] != df['TotalPrice']).sum())

Number of TotalPrice that are not calculated correctlly:  107


5.2 - Inspect the 107 mismatched records

In [49]:
mismatched = df[
    (df['Quantity'] * df['UnitPrice']).round(2) != df['TotalPrice'].round(2)
]

mismatched.head(20)

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice


In [50]:
mismatched[['Quantity', 'UnitPrice', 'CouponCode', 'TotalPrice']].head(20)

,Quantity,UnitPrice,CouponCode,TotalPrice


In [53]:
print((df['Quantity'] * df['UnitPrice']).round(2).eq(df['TotalPrice'].round(2)).all())

True


The `TotalPrice` column was validated against the expected calculation:

`Quantity × UnitPrice`

An initial direct comparison identified 107 apparent mismatches. However, after rounding both calculated and recorded values to two decimal places, all records matched.

This indicated that the initial differences were caused by floating-point precision rather than actual data inconsistencies.

The final validation returned `True`, confirming that all `TotalPrice` values were consistent with `Quantity × UnitPrice` to two decimal places.

No changes were made to the `TotalPrice` column.

6. Text Data Validation

6.1 - Check the text/categorical columns

In [54]:
text_columns = [
    'Product',
    'PaymentMethod',
    'OrderStatus',
    'CouponCode',
    'ReferralSource'
]

for col in text_columns:
    print(f"\n--- {col} ---")
    print(df[col].value_counts(dropna=False))


--- Product ---
Product
Printer    181
Tablet     179
Chair      178
Laptop     173
Desk       170
Monitor    163
Phone      156
Name: count, dtype: int64

--- PaymentMethod ---
PaymentMethod
Online         258
Cash           246
Credit Card    234
Debit Card     232
Gift Card      230
Name: count, dtype: int64

--- OrderStatus ---
OrderStatus
Cancelled    250
Returned     247
Pending      237
Shipped      235
Delivered    231
Name: count, dtype: int64

--- CouponCode ---
CouponCode
FREESHIP    313
Unknown     309
WINTER15    292
SAVE10      286
Name: count, dtype: int64

--- ReferralSource ---
ReferralSource
Instagram    259
Email        250
Google       241
Facebook     228
Referral     222
Name: count, dtype: int64


6.2 - Check for hidden spaces

In [55]:
for col in text_columns:
    spaces = df[col].astype(str).str.strip().ne(df[col].astype(str))
    print(f"{col}: {spaces.sum()} values with leading/trailing spaces")

Product: 0 values with leading/trailing spaces
PaymentMethod: 0 values with leading/trailing spaces
OrderStatus: 0 values with leading/trailing spaces
CouponCode: 0 values with leading/trailing spaces
ReferralSource: 0 values with leading/trailing spaces


Text Format Validation: The categorical columns were checked for inconsistent capitalization, spelling variations, and leading/trailing spaces. No formatting inconsistencies were identified, so no additional text transformations were required.

7. Identifier Validation

7.1 - Check the ID and identifier columns

In [56]:
#OrderID
print("Total OrderIDs:", df['OrderID'].nunique())
print("Total rows:", len(df))
print("Missing OrderIDs:", df['OrderID'].isnull().sum())

Total OrderIDs: 1200
Total rows: 1200
Missing OrderIDs: 0


In [57]:
#CustomerID
print("Unique CustomerIDs:", df['CustomerID'].nunique())
print("Total rows:", len(df))
print("Missing CustomerIDs:", df['CustomerID'].isnull().sum())

Unique CustomerIDs: 1189
Total rows: 1200
Missing CustomerIDs: 0


In [58]:
#TrackingNumber
print("Unique TrackingNumbers:", df['TrackingNumber'].nunique())
print("Total rows:", len(df))
print("Missing TrackingNumbers:", df['TrackingNumber'].isnull().sum())
print("Duplicate TrackingNumbers:", df['TrackingNumber'].duplicated().sum())

Unique TrackingNumbers: 1200
Total rows: 1200
Missing TrackingNumbers: 0
Duplicate TrackingNumbers: 0


8. Final Data Quality Validation

8.1 - Check for invalid or unusual categorical values

In [59]:
for col in ['Product', 'PaymentMethod', 'OrderStatus', 'ReferralSource']:
    print(f"\n{col}:")
    print(df[col].unique())


Product:
['Monitor' 'Phone' 'Tablet' 'Chair' 'Printer' 'Laptop' 'Desk']

PaymentMethod:
['Debit Card' 'Online' 'Credit Card' 'Gift Card' 'Cash']

OrderStatus:
['Shipped' 'Cancelled' 'Returned' 'Delivered' 'Pending']

ReferralSource:
['Instagram' 'Referral' 'Email' 'Facebook' 'Google']


8.2 - Check for missing values one final time

In [64]:
#Check for missing values in each column
df.isnull().sum()

,0
OrderID,0
Date,0
CustomerID,0
Product,0
Quantity,0
UnitPrice,0
ShippingAddress,0
PaymentMethod,0
OrderStatus,0
TrackingNumber,0


In [65]:
#Check for missing values in the entire dataset
print(df.isnull().sum().sum())

0


8.3 - Final duplicate validation

In [66]:
print("Duplicate rows:", df.duplicated().sum())
print("Duplicate OrderIDs:", df['OrderID'].duplicated().sum())

Duplicate rows: 0
Duplicate OrderIDs: 0


8.4 - Final data-quality summary

In [67]:
print("Dataset shape:", df.shape)
print("Total missing values:", df.isnull().sum().sum())
print("Duplicate rows:", df.duplicated().sum())
print("Duplicate OrderIDs:", df['OrderID'].duplicated().sum())
print("Duplicate TrackingNumbers:", df['TrackingNumber'].duplicated().sum())
print("Unique OrderIDs:", df['OrderID'].nunique())
print("Unique TrackingNumbers:", df['TrackingNumber'].nunique())
print("Date data type:", df['Date'].dtype)

Dataset shape: (1200, 14)
Total missing values: 0
Duplicate rows: 0
Duplicate OrderIDs: 0
Duplicate TrackingNumbers: 0
Unique OrderIDs: 1200
Unique TrackingNumbers: 1200
Date data type: datetime64[ns]


9. Save the cleaned dataset

In [68]:
df.to_excel("Cleaned_Dataset.xlsx", index=False)

In [71]:
#verifying if the file was created
import os
os.path.exists("Cleaned_Dataset.xlsx")

True